File: src/weaponsprocurement/gui/StockReviewState.java
package weaponsprocurement.gui;

import weaponsprocurement.core.StockCategory;
import weaponsprocurement.core.StockReviewConfig;
import weaponsprocurement.core.StockSortMode;
import weaponsprocurement.core.StockSourceMode;
import weaponsprocurement.core.StockItemType;

import java.util.Set;

public final class StockReviewState implements WimGuiScrollableListState {
    private final StockReviewExpansionState expansion;
    private final StockReviewFilterState filters;
    private final StockReviewSourceState source;
    private int listScrollOffset = 0;
    private String tradeWarning = "None";
    private float initialCredits = -1f;
    private float initialCargoCapacity = -1f;

    public StockReviewState(StockReviewConfig config) {
        this.expansion = new StockReviewExpansionState();
        this.filters = new StockReviewFilterState();
        this.source = new StockReviewSourceState(config);
    }

    public StockReviewState(StockReviewState source) {
        this.expansion = new StockReviewExpansionState(source.expansion);
        this.filters = new StockReviewFilterState(source.filters);
        this.source = new StockReviewSourceState(source.source);
        this.listScrollOffset = source.listScrollOffset;
        this.tradeWarning = source.tradeWarning;
        this.initialCredits = source.initialCredits;
        this.initialCargoCapacity = source.initialCargoCapacity;
    }

    public boolean isExpanded(StockCategory category) {
        return expansion.isExpanded(category);
    }

    public void toggle(StockCategory category) {
        expansion.toggle(category);
    }

    public boolean isExpanded(StockItemType itemType, StockCategory category) {
        return expansion.isExpanded(itemType, category);
    }

    public void toggle(StockItemType itemType, StockCategory category) {
        expansion.toggle(itemType, category);
    }

    public boolean isExpanded(StockItemType itemType) {
        return expansion.isExpanded(itemType);
    }

    public void toggle(StockItemType itemType) {
        expansion.toggle(itemType);
    }

    public boolean isExpanded(StockReviewTradeGroup tradeGroup) {
        return expansion.isExpanded(tradeGroup);
    }

    public void toggle(StockReviewTradeGroup tradeGroup) {
        expansion.toggle(tradeGroup);
    }

    public void setExpanded(StockReviewTradeGroup tradeGroup, boolean value) {
        expansion.setExpanded(tradeGroup, value);
    }

    public boolean isItemExpanded(String itemKey) {
        return expansion.isItemExpanded(itemKey);
    }

    public void toggleItem(String itemKey) {
        expansion.toggleItem(itemKey);
    }

    public boolean isFilterActive(StockReviewFilter filter) {
        return filters.isFilterActive(filter);
    }

    public void toggleFilter(StockReviewFilter filter) {
        filters.toggleFilter(filter);
        listScrollOffset = 0;
    }

    public Set<StockReviewFilter> getActiveFilters() {
        return filters.getActiveFilters();
    }

    public int getActiveFilterCount() {
        return filters.getActiveFilterCount();
    }

    public void clearFilters() {
        filters.clearFilters();
        listScrollOffset = 0;
    }

    public boolean isExpanded(StockReviewFilterGroup group) {
        return filters.isExpanded(group);
    }

    public void toggle(StockReviewFilterGroup group) {
        filters.toggle(group);
    }

    public StockSortMode getSortMode() {
        return source.getSortMode();
    }

    public void cycleSortMode() {
        source.cycleSortMode();
    }

    public boolean isIncludeCurrentMarketStorage() {
        return source.isIncludeCurrentMarketStorage();
    }

    public boolean isIncludeBlackMarket() {
        return source.isIncludeBlackMarket();
    }

    public void toggleBlackMarket() {
        source.toggleBlackMarket();
    }

    public StockSourceMode getSourceMode() {
        return source.getSourceMode();
    }

    public void cycleSourceMode() {
        source.cycleSourceMode();
    }

    public int getListScrollOffset() {
        return listScrollOffset;
    }

    @Override
    public void setListScrollOffset(int listScrollOffset) {
        this.listScrollOffset = Math.max(0, listScrollOffset);
    }

    public void adjustListScrollOffset(int delta, int maxOffset) {
        listScrollOffset = WimGuiScroll.usefulOffsetByDelta(listScrollOffset, delta, Math.max(0, maxOffset));
    }

    public String getTradeWarning() {
        return tradeWarning;
    }

    public void setTradeWarning(String tradeWarning) {
        this.tradeWarning = tradeWarning == null || tradeWarning.isEmpty() ? "None" : tradeWarning;
    }

    public float getInitialCredits() {
        return initialCredits;
    }

    public void setInitialCreditsIfUnset(float initialCredits) {
        if (this.initialCredits < 0f) {
            this.initialCredits = Math.max(0f, initialCredits);
        }
    }

    public float getInitialCargoCapacity() {
        return initialCargoCapacity;
    }

    public void setInitialCargoCapacityIfUnset(float initialCargoCapacity) {
        if (this.initialCargoCapacity < 0f) {
            this.initialCargoCapacity = Math.max(0f, initialCargoCapacity);
        }
    }

}

File: src/main/kotlin/weaponsprocurement/ui/stockreview/StockReviewState.kt
package weaponsprocurement.ui.stockreview

import weaponsprocurement.ui.*

import weaponsprocurement.stock.StockCategory
import weaponsprocurement.stock.StockItemType
import weaponsprocurement.config.StockReviewConfig
import weaponsprocurement.stock.StockSortMode
import weaponsprocurement.stock.StockSourceMode

class StockReviewState : WimGuiScrollableListState {
    private val expansion: StockReviewExpansionState
    private val filters: StockReviewFilterState
    private val source: StockReviewSourceState
    private var listScrollOffset = 0
    private var tradeWarning = "None"
    private var initialCredits = -1f
    private var initialCargoCapacity = -1f

    constructor(config: StockReviewConfig) {
        expansion = StockReviewExpansionState()
        filters = StockReviewFilterState()
        source = StockReviewSourceState(config)
    }

    constructor(source: StockReviewState) {
        expansion = StockReviewExpansionState(source.expansion)
        filters = StockReviewFilterState(source.filters)
        this.source = StockReviewSourceState(source.source)
        listScrollOffset = source.listScrollOffset
        tradeWarning = source.tradeWarning
        initialCredits = source.initialCredits
        initialCargoCapacity = source.initialCargoCapacity
    }

    fun isExpanded(category: StockCategory?): Boolean = expansion.isExpanded(category)
    fun toggle(category: StockCategory?) = expansion.toggle(category)
    fun isExpanded(itemType: StockItemType?, category: StockCategory?): Boolean = expansion.isExpanded(itemType, category)
    fun toggle(itemType: StockItemType?, category: StockCategory?) = expansion.toggle(itemType, category)
    fun isExpanded(itemType: StockItemType?): Boolean = expansion.isExpanded(itemType)
    fun toggle(itemType: StockItemType?) = expansion.toggle(itemType)
    fun isExpanded(tradeGroup: StockReviewTradeGroup?): Boolean = expansion.isExpanded(tradeGroup)
    fun toggle(tradeGroup: StockReviewTradeGroup?) = expansion.toggle(tradeGroup)
    fun setExpanded(tradeGroup: StockReviewTradeGroup?, value: Boolean) = expansion.setExpanded(tradeGroup, value)
    fun isItemExpanded(itemKey: String?): Boolean = expansion.isItemExpanded(itemKey)
    fun toggleItem(itemKey: String?) = expansion.toggleItem(itemKey)

    fun isFilterActive(filter: StockReviewFilter?): Boolean = filters.isFilterActive(filter)

    fun toggleFilter(filter: StockReviewFilter?) {
        filters.toggleFilter(filter)
        listScrollOffset = 0
    }

    fun getActiveFilters(): Set<StockReviewFilter> = filters.getActiveFilters()
    fun getActiveFilterCount(): Int = filters.getActiveFilterCount()

    fun clearFilters() {
        filters.clearFilters()
        listScrollOffset = 0
    }

    fun isExpanded(group: StockReviewFilterGroup?): Boolean = filters.isExpanded(group)
    fun toggle(group: StockReviewFilterGroup?) = filters.toggle(group)
    fun getSortMode(): StockSortMode = source.getSortMode()
    fun cycleSortMode() = source.cycleSortMode()
    fun isIncludeCurrentMarketStorage(): Boolean = source.isIncludeCurrentMarketStorage()
    fun isIncludeBlackMarket(): Boolean = source.isIncludeBlackMarket()
    fun toggleBlackMarket() = source.toggleBlackMarket()
    fun getSourceMode(): StockSourceMode = source.getSourceMode()
    fun cycleSourceMode() = source.cycleSourceMode()

    override fun getListScrollOffset(): Int = listScrollOffset

    override fun setListScrollOffset(listScrollOffset: Int) {
        this.listScrollOffset = Math.max(0, listScrollOffset)
    }

    fun adjustListScrollOffset(delta: Int, maxOffset: Int) {
        listScrollOffset = WimGuiScroll.usefulOffsetByDelta(listScrollOffset, delta, Math.max(0, maxOffset))
    }

    fun getTradeWarning(): String = tradeWarning

    fun setTradeWarning(tradeWarning: String?) {
        this.tradeWarning = if (tradeWarning.isNullOrEmpty()) "None" else tradeWarning
    }

    fun getInitialCredits(): Float = initialCredits

    fun setInitialCreditsIfUnset(initialCredits: Float) {
        if (this.initialCredits < 0f) {
            this.initialCredits = Math.max(0f, initialCredits)
        }
    }

    fun getInitialCargoCapacity(): Float = initialCargoCapacity

    fun setInitialCargoCapacityIfUnset(initialCargoCapacity: Float) {
        if (this.initialCargoCapacity < 0f) {
            this.initialCargoCapacity = Math.max(0f, initialCargoCapacity)
        }
    }
}
